### PR TITLE
LF: fix specification of exceptions

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This is part of #8020.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/spec/daml-lf-1.rst [hunk 1]
```diff
@@ -1144,13 +1145,19 @@ Then we define *well-formed expressions*. ::
...
       Γ  ⊢  eₘ  : 'Text'
       Γ  ⊢  eₚ  :  τ
     ——————————————————————————————————————————————————————————————— ExpMakeAnyException [DAML-LF ≥ 1.dev]
       Γ  ⊢  'make_any_exception' @τ eₘ eₚ  :  'AnyException'
 
-      ε  ⊢  τ  :  ⋆      ⊢ₑ  τ
+      ⊢ₑ  τ
       Γ  ⊢  e  :  'AnyException'
     ——————————————————————————————————————————————————————————————— ExpFromAnyException [DAML-LF ≥ 1.dev]
       Γ  ⊢  'from_any_exception' @τ e  :  'Option' τ
```
Comment (sofiafaro-da) [inline daml-lf/spec/daml-lf-1.rst:1163]:
```suggestion
      Γ  ⊢  'from_any_exception' @τ e  :  'Optional' τ
```

File: daml-lf/spec/daml-lf-1.rst [hunk 2]
```diff
@@ -2420,7 +2427,7 @@ exact output.
 
       e  ⇓  Ok v
     —————————————————————————————————————————————————————————————————————— EvExpThrow
-      'THROW' @τ e  ⇓  Err v
+      'throw' @τ e  ⇓  Err v
```
Comment (sofiafaro-da) [inline daml-lf/spec/daml-lf-1.rst:2430]: Takes an additional type parameter:
```suggestion
      'throw' @σ @τ e  ⇓  Err v
```

File: daml-lf/spec/daml-lf-1.rst [hunk 3]
```diff
@@ -4126,14 +4133,7 @@ Error functions
...
         Λ (α : ⋆). λ (x : 'Text').
-        'THROW' @α ('make_any_exception' @'GeneralError' x ('MAKE_GENERAL_ERROR' x))
-
-* ``THROW : ∀ (α : ⋆) . 'AnyException' → α``
-
-  [*Available in version >= 1.11*]
-
-  Throws an ``'AnyException'``. See the evaluation rule ``EvExpThrow`` for
-  precise semantics.
+        'throw' @α ('make_any_exception' @'GeneralError' x ('MAKE_GENERAL_ERROR' x))
```
Comment (sofiafaro-da) [inline daml-lf/spec/daml-lf-1.rst:4136]: 'throw' now throws exceptions directly:
```suggestion
        'throw' @α @'GeneralError' ('MAKE_GENERAL_ERROR' x)
```

File: daml-lf/spec/daml-lf-1.rst [hunk 4]
```diff
@@ -1144,13 +1145,19 @@ Then we define *well-formed expressions*. ::
     ——————————————————————————————————————————————————————————————— ExpCase
       Γ  ⊢  'case' e 'of' alt₁ | … | altₙ : σ
 
-      ε  ⊢  τ  :  ⋆      ⊢ₑ  τ
+      Γ  ⊢  σ  :  ⋆
+      ⊢ₑ  τ
+      Γ  ⊢  e  :  σ
+    ——————————————————————————————————————————————————————————————— ExpThrow [DAML-LF ≥ 1.dev]
+      Γ  ⊢  'throw' @σ  @τ @e  :  τ
```
Comment (sofiafaro-da) [inline daml-lf/spec/daml-lf-1.rst:1152]: The return types & exception type were mixed up:

```suggestion
      Γ  ⊢  σ  :  ⋆
      ⊢ₑ  τ
      Γ  ⊢  e  :  τ
    ——————————————————————————————————————————————————————————————— ExpThrow [DAML-LF ≥ 1.dev]
      Γ  ⊢  'throw' @σ  @τ @e  :  σ
```

**Review by sofiafaro-da — APPROVED** [review]
Thanks!

File: daml-lf/spec/daml-lf-1.rst [hunk 5]
```diff
@@ -603,6 +603,7 @@ Then we can define our kinds, types, and expressions::
        | 'type_rep' @τ                              -- ExpToTypeRep: A type representation [DAML-LF ≥ 1.7]
        |  u                                         -- ExpUpdate: Update expression
        |  s                                         -- ExpScenario: Scenario expression
+       | 'throw'                                    -- ExpThrow: throw exception
```
Comment (hurryabit) [inline daml-lf/spec/daml-lf-1.rst:606]: Shouldn't this be something like
```suggestion
       | 'throw' @σ @τ e                            -- ExpThrow: throw exception
```

File: daml-lf/spec/daml-lf-1.rst [hunk 6]
```diff
@@ -702,7 +703,7 @@ available for usage::
             , 'choices' { ChDef₁, …, ChDefₘ }
             , KeyDef
             }
-       |  'exception' (x: T)                        -- DefException [DAML-LF ≥ 1.dev]
+       |  'exception' T                             -- DefException [DAML-LF ≥ 1.dev]
```
Comment (hurryabit) [inline daml-lf/spec/daml-lf-1.rst:706]: We need to add the `message` here as well, don't we?

**Review by hurryabit — COMMENTED** [review]
Good catches. Thanks.

File: daml-lf/spec/daml-lf-1.rst (hunk 6 above)
Comment (remyhaemmerle-da) [reply daml-lf/spec/daml-lf-1.rst:706]: Indeed.

File: daml-lf/spec/daml-lf-1.rst (hunk 5 above)
Comment (remyhaemmerle-da) [reply daml-lf/spec/daml-lf-1.rst:606]: Good catch.

**Pull request merged**